Adding a registers.sw module

nfurfaro opened the pull request:
Adds convenience functions for the 16 reserved VM registers to avoid the need to use `asm` blocks to access them multiple times in the stdlib, or by users developing other libraries.

This addresses some of https://github.com/FuelLabs/sway-lib-std/issues/38.
A follow-up task will be to refactor the stdlib to leverage these functions instead of accessing registers directly, tracked here: FuelLabs/sway#1079

This should probably have some tests, but is also quite simple and directly follows the spec...

nfurfaro committed:
feat:add bare module registers

nfurfaro committed:
feat:add registers.sw to expose16 reserved registers

nfurfaro commented on src/registers.sw:
```diff
@@ -0,0 +1,115 @@
+library registers;
+
+/// Contains zero (0), for convenience.
+pub fn zero() -> u64 {
+    asm() {
+        zero
+    }
+}
+
+/// Contains one (1), for convenience.
+pub fn one() -> u64 {
+    asm() {
+        one
+    }
+}
```
Added for completeness, but these don't add any value outside of `asm` blocks, so should probably be cut.

nfurfaro committed:
test:add new test project "registers"

nfurfaro committed:
build:update forc manifest file

nfurfaro committed:
test:add contract and abi.json

nfurfaro committed:
test:add test file and empty tests

nfurfaro commented:
Update: Adding some tests for this now.

nfurfaro committed:
test:temp disable ret & retl, add simple tests

otrho reviewed (changes requested):
Yeah, I agree -- remove the `zero` and `one` functions and tests and then it's all good.

adlerjohn commented on src/registers.sw:
```diff
@@ -0,0 +1,115 @@
+    }
+}
+
+/// Return value or pointer.
+// pub fn return_value() -> u64 {
+//     asm() {
+//         ret
+//     }
+// }
+
+/// Return value length in bytes.
+// pub fn return_length() -> u64 {
+//     asm() {
+//         retl
+//     }
+// }
```
Is there a reason these are commented out?

nfurfaro replied:
I'm investigating. It's because of this error:
```
error
  --> sway-lib-std/src/registers.sw:86:9
   |
84 |
85 |     asm() {
86 |         ret
   |         ^^^ This register was not initialized in the initialization section of the ASM expression. Initialized registers are:
87 |     }
88 | }
   |
error
  --> sway-lib-std/src/registers.sw:93:9
   |
91 |
92 |     asm() {
93 |         retl
   |         ^^^^ This register was not initialized in the initialization section of the ASM expression. Initialized registers are:
94 |     }
95 | }
   |
  Aborting due to 2 errors.
Error: "Failed to compile registers"
```

nfurfaro replied:
It seems that `ret` and `retl` are not recognized as named registers.

mohammadfawaz replied:
Looks like we're expecting a different spelling in the compiler: https://github.com/FuelLabs/sway/blob/916123ecda8a8a222fad63098ab36583871b2a13/sway-core/src/asm_lang/mod.rs#L1069
Should be an easy fix.

nfurfaro replied:
oh yeah. 
Do we stick to the spec and rename in the compiler, or edit the spec and change them here?

mohammadfawaz replied:
I think the spec is the reference we should follow. I can create a quick PR to update the compiler.

adlerjohn reviewed (commented):
Wait, can we re-add `zero` and `one`? The reason is that I'm thinking that the optimizer can remove the use of temporary registers, and use `$zero` or `$one` directly in instructions _that don't take an immediate_. The most obvious example being conditional jumps.

adlerjohn commented:
Converting to draft while waiting for a new release of `forc`.

nfurfaro commented:
@adlerjohn Can the optimizer just use `asm` to access `$zero` and `$one` directly, without the need for wrapper functions?

mohammadfawaz commented:
> @adlerjohn Can the optimizer just use `asm` to access `$zero` and `$one` directly, without the need for wrapper functions?

I think it should - it would be a bit weird if using `one()` instead of `1` leads to different bytecode, right?

adlerjohn commented:
> Can the optimizer just use asm to access $zero and $one directly, without the need for wrapper functions?

The optimizer can inline the call to the `zero()` function, which assigns `$zero` to some register. Then since that register isn't written to later on, the optimizer can remove it and use `$zero` directly where that register was being read/used.

E.g. it could transform

```
move $rA $zero
add $rB $rC $rA
```

into

```
add $rB $rC $zero
```

(note that `ADD` is used here and not `ADDI`, so the compiler can't replace `$zero` with an immediate `0`)

(double-note: a super-clever optimizer could replace the `ADD` with an `ADDI` in this trivial example, but you get the idea 😂)

nfurfaro committed:
test:enable ret & retl again

otrho commented:
There is no reason for a developer writing Sway to use `one()` instead of a literal `1`.  The compiler will automatically work out when it's better to use `$one` instead of loading a literal from memory, this isn't something the libraries should worry about.

adlerjohn commented:
> There is no reason for a developer writing Sway to use `one()` instead of a literal `1`

That's actually a very good point. But until the compiler can do that, is it not better to have a `one()` and `zero()` available for the standard library to use, in places immediates aren't allowed?

otrho commented:
Immediates are allowed everywhere... :slightly_smiling_face:    `$one` and `$zero` are still definitely very handy within ASM blocks, but we're not suggesting changing that.  The point of this PR is just to allow developers to use e.g., `program_counter()` instead of `asm() { pc }`.  It's almost just syntactic sugar, with better typing and verbose naming.

> But until the compiler can do that

The compiler _can_ do that, it's just a part of ASM gen.  At the moment we still load all immediates from the data section but this can and will change with IR.  Having even the core library developers think in terms of such low level optimisations isn't sustainable, let alone regular developers.

nfurfaro committed:
Merge branch 'master' into furnic/registers-module

nfurfaro committed:
build:update manifest authors field

nfurfaro committed:
build:remove redundant abi file

nfurfaro committed:
fix:remove unneeded zero and one functions

nfurfaro committed:
fix:remove zero and one funcs from abi and contract

nfurfaro committed:
test:remove tests for zero and one & updates paths to abi file

nfurfaro committed:
Merge branch 'master' into furnic/registers-module

nfurfaro committed:
test:fix expected values and function calls

nfurfaro committed:
style:forc fmt

nfurfaro commented:
I believe I've addressed all the comments here, but I can't merge until a new version of fuels-rs is released (due to the fix for the forc.toml file "authors" field only existing on fules-rs master atm).
Leaving in draft status till resolved.

nfurfaro committed:
fix: add dep registers to context.sw

nfurfaro committed:
style: forc fmt

nfurfaro committed:
build:bump sdk version and add fuels-signers as a dep

nfurfaro committed:
update tests to use new syntax

nfurfaro committed:
Merge branch 'master' into furnic/registers-module

nfurfaro committed:
style:cargo fmt & forc fmt
